build-scripts/build_tests.py

```diff
@@ -0,0 +1,230 @@
+            content = ssg.jinja.process_file_with_macros(str(test.absolute()), env_yaml)
+            rule_output_path.mkdir(parents=True, exist_ok=True)
+            _write_path(content, rule_output_path / test.name)
+
+
+def _process_rules(benchmark_cpes: Set[str], env_yaml: Dict, output_path: pathlib.Path,
+                   rules: Iterable[pathlib.Path], templates_root: pathlib.Path,
+                   product_rules: list) -> None:
+    logger = logging.getLogger()
+    for rule_file in rules:  # type: pathlib.Path
+        rendered_rule_obj = ssg.yaml.open_raw(str(rule_file))
+        rule_path = pathlib.Path(rendered_rule_obj["definition_location"])
+        rule_root = rule_path.parent
+        rule_id = rule_root.name
+        if rule_id not in product_rules:
+            logger.debug("Skipping %s since it is not in the product", rule_id)
+            continue
+        rule_tests_root = rule_root / "tests"
+        rule_output_path = output_path / rule_id
+
+        if rule_tests_root.exists():
+            _process_local_tests(benchmark_cpes, env_yaml, rule_output_path, rule_tests_root)
+        if rendered_rule_obj["template"] is not None:
+            if "name" not in rendered_rule_obj["template"]:
+                raise ValueError(f"Invalid template config on rule {rule_id}")
+            template_name = rendered_rule_obj["template"]["name"]
+            template_root = templates_root / template_name
+            template_tests_root = template_root / "tests"
+            if not template_tests_root.exists():
+                logger.debug("Template %s doesn't have tests. Skipping for rule %s.",
+                             template_name, rule_id)
+                continue
+            test_config_path = rule_tests_root / "test_config.yml"
+            deny_templated_scenarios = _get_deny_templated_scenarios(test_config_path)
+            for test in template_tests_root.iterdir():  # type: pathlib.Path
+                if not test.name.endswith(".sh") or test.name in deny_templated_scenarios:
+                    logging.warning("Skipping % for %s", test.name, rule_id)
+                    continue
+                template = ssg.templates.Template.load_template(str(templates_root.absolute()),
+                                                                template_name)
+                rendered_rule_obj["template"]["vars"]["_rule_id"] = rule_id
+                template_parameters = template.preprocess(
+                    rendered_rule_obj["template"]["vars"], "test")
+                env_yaml = env_yaml.copy()
+                jinja_dict = ssg.utils.merge_dicts(env_yaml, template_parameters)
+                file_contents = ssg.jinja.process_file_with_macros(str(test.absolute()),
+                                                                   jinja_dict)
+                scenario = tests.ssg_test_suite.rule.Scenario(test.name, file_contents)
+                if scenario.matches_platform(benchmark_cpes):
```

Comment:
This is very inefficient.

The method "matches_platform" loads all CPEs for all platforms. It calls `common.matches_platform` which calls `common._get_platform_cpes` which iterates over all matched products and loads all CPEs for these products.

Therefore for many test scenarios we load many CPEs, including CPEs for other products and we repeat the loading for each test scenario. The result is that this script spends most of the time in loading CPEs.

We need to create a different and simplified way for test scenarios platform matching here.